cmake/templates/HPXMacros.cmake.in

```diff
@@ -11,13 +11,41 @@ include(GNUInstallDirs)
 include(HPX_Utils)
 
 function(hpx_check_compiler_compatibility)
-    if(HPX_IGNORE_COMPILER_COMPATIBILITY OR HPX_WITH_BUILD_BINARY_PACKAGE)
+    if(HPX_IGNORE_COMPILER_COMPATIBILITY)
     return()
-  endif()
-
-  if(NOT (${CMAKE_CXX_COMPILER_ID} STREQUAL ${HPX_CXX_COMPILER_ID}
-     AND ${CMAKE_CXX_COMPILER_VERSION} STREQUAL ${HPX_CXX_COMPILER_VERSION}))
-    set(MESSAGE "Compilers do not match. In order to compile HPX application it is"
+    endif()
+       
+    if(HPX_WITH_BUILD_BINARY_PACKAGE)
+    
+    string(REGEX MATCH "[0-9]+" HPX_MAJOR_VERSION "${HPX_CXX_COMPILER_VERSION}")
+    string(REGEX MATCH "[0-9]+" CXX_MAJOR_VERSION "${CMAKE_CXX_COMPILER_VERSION}")
+   
+        if( NOT (${HPX_MAJOR_VERSION} STREQUAL ${CXX_MAJOR_VERSION}) )
+        set(MESSAGE "Compilers do not match. In order to compile HPX application it is"
+        " recommended to use the same major version of the compiler as you did for HPX. "
+        "HPX_CXX_COMPILER=${HPX_CXX_COMPILER}, "
+        "CMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}. "
+        "HPX_CXX_COMPILER_ID=${HPX_CXX_COMPILER_ID}, "
+        "CMAKE_CXX_COMPILER_ID=${CMAKE_CXX_COMPILER_ID}. "
+        "HPX_CXX_MAJOR_VERSION=${HPX_MAJOR_VERSION}, "
+        "CMAKE_CXX_COMPILER_VERSION=${CXX_MAJOR_VERSION}."
```

Comment:
This text is inconsistent, please make the generated output understandable.

Reply:
Done